File: C++/minimum-number-of-operations-to-sort-a-binary-tree-by-level.cpp
// Time:  O(nlogn)
// Space: O(w)

// bfs, sort
class Solution {
public:
    int minimumOperations(TreeNode* root) {
        int result = 0;
        vector<TreeNode *> q = {root};
        while (!empty(q)) {
            vector<TreeNode *> new_q;
            for (const auto& node : q) {
                if (node->left) {
                    new_q.emplace_back(node->left);
                }
                if (node->right) {
                    new_q.emplace_back(node->right);
                }
            }
            vector<int> idx(size(q));
            iota(begin(idx), end(idx), 0);
            sort(begin(idx), end(idx), [&](int a, int b) { return q[a]->val < q[b]->val; });
            for (int i = 0; i < size(idx); ++i) {
                for (; idx[i] != i; swap(idx[idx[i]], idx[i]), ++result);
            }
            q = move(new_q);
        }
        return result;
    }
};

File: Python/minimum-number-of-operations-to-sort-a-binary-tree-by-level.py
# Time:  O(nlogn)
# Space: O(w)

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        pass


# bfs, sort
class Solution(object):
    def minimumOperations(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        result = 0
        q = [root]
        while q:
            new_q = []
            for node in q:
                if node.left:
                    new_q.append(node.left)
                if node.right:
                    new_q.append(node.right)
            idx = range(len(q))
            idx.sort(key=lambda x: q[x].val)
            for i in xrange(len(q)):
                while idx[i] != i:
                    idx[idx[i]], idx[i] = idx[i], idx[idx[i]]
                    result += 1
            q = new_q
        return result
